### DISCUSSION TITLE
Change Keys for flash.nvim

### DISCUSSION THREAD

**Discussion by sagarPakhrin** [LazyVim/LazyVim#1299] [General]
I like the nvim flash plugin that comes with lazy vim starter. However, I like what vim does by default with "s" and "S" keys.

This is what comes with lazy starter
```lua
{
    "folke/flash.nvim",
    event = "VeryLazy",
    vscode = true,
    ---@type Flash.Config
    opts = {},
    -- stylua: ignore
    keys = {
      { "s", mode = { "n", "x", "o" }, function() require("flash").jump() end, desc = "Flash" },
      { "S", mode = { "n", "o", "x" }, function() require("flash").treesitter() end, desc = "Flash Treesitter" },
      { "r", mode = "o", function() require("flash").remote() end, desc = "Remote Flash" },
      { "R", mode = { "o", "x" }, function() require("flash").treesitter_search() end, desc = "Treesitter Search" },
      { "<c-s>", mode = { "c" }, function() require("flash").toggle() end, desc = "Toggle Flash Search" },
    },
  },
```
I created a flash.lua file inside `plugins` directory and have this config.
I tried configuring the flash to different key (`\s`) but it seems to merge the two and not replace.
How do i revert the default behavior of "s" and "S" and have a diferent key to trigger flash
```lua
{
    "folke/flash.nvim",
    event = "VeryLazy",
    vscode = true,
    ---@type Flash.Config
    opts = {},
    -- stylua: ignore
    keys = {
      { "\\s", mode = { "n", "x", "o" }, function() require("flash").jump() end, desc = "Flash" },
      { "\\SS", mode = { "n", "o", "x" }, function() require("flash").treesitter() end, desc = "Flash Treesitter" },
      { "r", mode = "o", function() require("flash").remote() end, desc = "Remote Flash" },
      { "R", mode = { "o", "x" }, function() require("flash").treesitter_search() end, desc = "Treesitter Search" },
      { "<c-s>", mode = { "c" }, function() require("flash").toggle() end, desc = "Toggle Flash Search" },
    },
  },
```

**Comment by dpetka2001** [comment]
You have to explicitly disable the keymaps you want for all the modes they were declared to begin with. Because your own spec will merge with the default one and you will end up with 2 different keymaps bound to the same action. Also, if you don't want to change the rest of the keys `r/R/<c-s>` you don't have to declare them in your own spec.

**Reply by swelljoe** [reply]
Can you provide an example? I've tried to do that, but when I add this to my lua based on my understanding of [this doc](https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps):

```
{ "folke/flash.nvim", keys = { "s", false } },
```

I get this error on nvim start:

```
Error detected while processing User Autocommands for "VeryLazy":
Error executing lua callback: ...share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/keys.lua:22: attempt to index local 'ret' (a boolean value)
stack traceback:
        ...share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/keys.lua:22: in function 'parse'
        ...share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/keys.lua:44: in function 'values'
        ...share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/init.lua:103: in function 'del'
        ...share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/init.lua:43: in function 'disable'
        ...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:325: in function '_load'
        ...local/share/nvim/lazy/lazy.nvim/lua/lazy/core/loader.lua:185: in function 'load'
        ...hare/nvim/lazy/lazy.nvim/lua/lazy/core/handler/event.lua:31: in function <...hare/nvim/lazy/lazy.nvim/lua/lazy/core/handler/event.lua:23>
        [C]: in function 'nvim_exec_autocmds'
        .../.local/share/nvim/lazy/lazy.nvim/lua/lazy/core/util.lua:149: in function <.../.local/share/nvim/lazy/lazy.nvim/lua/lazy/core/util.lua:144>
```

And, seems to completely disable all `flash` keymaps.

**Reply by folke** [reply]
The syntax you're using is completely wrong. Please read the lazy.nvim docs.

**Reply by swelljoe** [reply]
I'm trying. I don't understand them. The doc I'm looking at is this: https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps

I don't understand what I'm missing.

**Reply by dpetka2001** [reply]
You can see [here](https://www.lazyvim.org/plugins/editor#flashnvim) how `flash` is configured. `s` is configured for `mode = { "n", "x", "o" }`. So you would have to add `mode = { "n", "x", "o" }` to your example for disabling the `s` key.

**Reply by dpetka2001** [reply]
Also you need another set of curly braces right after the `keys` section. Like this
```lua
    keys = {
      { "s", mode = { "n", "x", "o" }, false },
	},
```

**Comment by git-rz** [comment]
The choice to override the default behavior of `s`, `S`, `r`, `R`, has caused me significant pain. It is a blight on an otherwise superb packaging of nvim. Please consider moving these to `<leader>j`, `<leader>t`, `<leader>r`, `<leader>T`.

Also, it would be very helpful for the `lua/plugins/example.lua` file to have an example that removes a key binding:

```
 return {
  {
    "folke/flash.nvim", enabled = true,
    keys= {
      { "s", mode = { "n", "x", "o" }, false },
      { "S", mode = { "n", "o", "x" }, false },
      { "r", mode = "o", false },
      { "R", mode = { "o", "x" }, false },
      { "<leader>j", mode = { "n", "x", "o" }, function() require("flash").jump() end, desc = "Flash" },
      { "<leader>t", mode = { "n", "o", "x" }, function() require("flash").treesitter() end, desc = "Flash Treesitter" },
      { "<leader>r", mode = "o", function() require("flash").remote() end, desc = "Remote Flash" },
      { "<leader>T", mode = { "o", "x" }, function() require("flash").treesitter_search() end, desc = "Treesitter Search" },
    }
  }
```

**Reply by dpetka2001** [reply]
A "blight" is a bit of overreacting in my opinion. You can change them to whatever you like in your personal config (just like i did with mine, but don't go around complaining about defaults). No defaults will ever satisfy everyone and that's why most ppl tinker their personal config to their needs. Also, there's already example in the [docs](https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps) about changing/disabling plugin keymaps.

**Reply by git-rz** [reply]
LazyVim bills itself as having sane defaults. I'm humbly submitting that these few choices do not fit my definition of that criterion, for the following reasons:

- `R`: the premiere way to enter REPLACE mode. REPLACE mode being one of a very few modes in what is a modal editor. (Mac users don't have an `insert` key to switch modes) 
- `s`: used (by me) almost as much as `a`, `o`, `O`, or `i`. Remapping any of those would also not be a sane default.
- `r` at least this isn't a major mode-switching workhorse like `R` or `s`, but still very common for me personally. Without `r`, I would use `s<new-char><esc>`, or `R<new-char><esc>`, but... That leaves `xi<new-char><esc>` which is not dot-repeatable. 
- `S`: this one is probably fine; most people probably `cc` anyway.

I'd love to see telemetry on most used vim keystrokes. Maybe I am an extreme outlier, and no one really uses `s` or `R`, and these choices are really just preference. But I'd be willing to bet that restoring these particular mappings will be the very first order of business for many people new to LazyVim, and who decide to continue with it despite this odd choice. I almost didn't.

I provide this feedback in the hope that more feedback might be collected and, if the data bears it out, the defaults might be changed. Short of that, the code snippet above could be added to the `example.lua` to simplify the process for the new users like OP and me.

**Reply by dpetka2001** [reply]
`r` is only used in operation pending mode. Likewise `R` is used in operation and visual mode. You can still use both in normal mode to do the normal vim operations. The only exception would be `s`, which you can remap or use `cl` instead (one more keystroke for essentially the same thing). I can totally understand your opinion, but i've seen many plugins proclaim sane defaults and I honestly don't pay attention to such expressions, because i don't believe there ever will be sane defaults for everyone (but i do understand it from a pure vim point of view regarding operations, it's just that it's not that big of a deal to me cuz i'll do my own customizations if something doesn't align with my workflow). As for the code snippet, that literally exists in the docs in the link i provided for how to disable `s` for flash plugin. But feel free to submit a PR if you think it should be included in the `example.lua` as well.

**Reply by LintaoAmons** [reply]
I use `s/S` for surrounding plugins. This default keybinding cause some pain and spend some time to try to totally disable it~ 
It would be good to have this written in the doc to show people how to disable the default keybindings

**Reply by dpetka2001** [reply]
Hey @LintaoAmons, did you take a look at the [docs](https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-adding--disabling-plugin-keymaps)? There's an example specifically for flash disabling the `s` key. I don't recall if it was present when the OP opened this topic, but it certainly got added because there were many topics like this one. Plz take a moment to read the docs to see if what you're searching for is in there.

**Reply by git-rz** [reply]
Hey @dpetka2001 , thanks for clarifying the scope of the `r` and `R` bindings. I am obviously new to LazyVim, and I was wrong about those binds. Honestly, I was so frustrated with `s`, that I just immediately disabled the entire plugin. I only reenabled it because I've never used a flash-style plugin, and they seem to be too popular to not be useful. I'll restore the `r` and `R` binds now, and try it out some more.

**Reply by dpetka2001** [reply]
No problem, I myself have also remapped `s` itself, but because of conflict with another plugin. I change whatever I find becomes uncomfortable for me. That's why I don't care about defaults or whatever. `flash` itself is pretty useful as a plugin, for me at least, so I would encourage that it's worth trying out and if you don't like it, just ditch it. Simple as that. Whatever makes you more comfortable.

**Comment by mmottl** [comment]
It seems the `s/S` flash bindings are well-chosen since they can both be replicated with reasonably quick and equivalent standard key strokes:

- `s` -> `cl`
- `S` -> `cc`

I haven't found any more convenient alternative. Just retrain your muscle memory!

**Reply by adursh** [reply]
Reason behind cl and cc?

**Reply by mmottl** [reply]
> Reason behind cl and cc? 

`c` clears using the subsequent movement `l` (next character to the right). `cc` clears the current line. Both are standard Vim bindings and should always work without having to rebind anything.

**Reply by bulletmark** [reply]
Note that `cc` and `cl` are standard mappings that have been in `vi` for more than 40 years (as have their `s` and `S` equivalents).

Saying that, I have been using `vi/vim` for almost that long so trying to change my reflex behaviour from `s` -> `cl` and `S` -> `cc` and changing to use flash search is a struggle!
